### ISSUE TITLE
After `ua attach` failure, detaching is required

### ISSUE THREAD

**Issue by frankban** [canonical/ubuntu-advantage-client#1337]
Related to https://github.com/canonical/ubuntu-advantage-client/issues/1335

In cases like the one above, from the customer perspective, `ua attach tokjen1` fails, then I try to attach token2 and it fails again, because the machine is already attached.

A `ua attach` failure should not leave the machine in a partial buggy state.

**Comment by TeodorPt** [comment]
Additional repro steps:

Because of #1335 attaching a new machine failed, with message:
```
Subscription has expired
To obtain a token please visit: https://ubuntu.com/advantage
```
However, ‘ua attach’ again with a different contract token, it says:
```
This machine is already attached to <>
To use a different subscription first run: sudo ua detach 
```
Checking the backend, the new machine is actually attached to the original contract (of the first command).

It seems the attaching actually works behind the scenes, just from the customer perspective it seems like it does not.

**Comment by TeodorPt** [comment]
Additional repro steps:

Because of #1335 attaching a new machine failed, with message:
```
Subscription has expired
To obtain a token please visit: https://ubuntu.com/advantage
```
However, ‘ua attach’ again with a different contract token, it says:
```
This machine is already attached to <>
To use a different subscription first run: sudo ua detach 
```

It looks like the machine is attached to the original token (of the first command), but from the customer perspective it seems like it is not. There might be similar situations.

**Comment by blackboxsw** [comment]
Thanks for the issue and content @frankban and @TeodorPt 

TLDR: invalid token mentioned in the description is a non-issue because we get a 401 from the contract server anyway. But, I'll confirm that other failure paths during a failed "initial attach" perform the appropriate cleanup of an unprocessed machine-token. 

Thankfully the PR @lucasmoura #1336 removed this raise on contract expiry which was one of few paths which lead to leaving an unread machine-token on disk during a failed initial machine attach. So that particular cause is not longer present in uaclient.  


> Additional repro steps:
> 
> Because of #1335 attaching a new machine failed, with message:
> 
> ```
> Subscription has expired
> To obtain a token please visit: https://ubuntu.com/advantage
> ```
> 
> However, ‘ua attach’ again with a different contract token, it says:
> 
> ```
> This machine is already attached to <>
> To use a different subscription first run: sudo ua detach 
> ```
> 
> It looks like the machine is attached to the original token (of the first command), but from the customer perspective it seems like it is not. There might be similar situations.

Per @frankban's description above. We don't have a specific ability to get into this 'false atteched' state due to an invalid token because the contract service routes return HTTP CODE 401 in those cases and no machine-token is presented to the client.

> 
> LE: another observation:
> When the above error is given by the ua client, the services will not be enabled automatically. However, upon enabling with `ua enable` it works fine.

It was an oversight that the UA Client enforced contract expiration it generally should rely on either service endpoints (authenticated APT repos or contract API interactions) to reject continued use/abuse of contract entitlements. 

Since expiry determination required that a valid contract were already obtained on disk, the client then sees itself as  "attached" and since the client wasn't supposed to enforce contract expiry for any operations it follows that `ua enable/disable X` would all just work.

**Comment by blackboxsw** [comment]
I believe  we have closed the potentially failure paths here that will result in a completely detached system if we fail during initial/partial attach. We will re-open this if we end up introducing paths that regress this behavior
